framework/include/timeintegrators/NonlinearTimeIntegratorInterface.h

```diff
@@ -0,0 +1,93 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+// Moose includes
+#include "MooseTypes.h"
+
+// Libmesh includes
+#include "libmesh/enum_parallel_type.h"
+
+// Forward declarations
+class SystemBase;
+class FEProblemBase;
+class NonlinearSystemBase;
+
+namespace libMesh
+{
+template <typename T>
+class NumericVector;
+class NonlinearImplicitSystem;
+} // namespace libMesh
+
+/**
+ * Interface class for routines and member variables for time integrators
+ * relying on Newton's method.
+ */
+class NonlinearTimeIntegratorInterface
+{
+public:
+  NonlinearTimeIntegratorInterface(FEProblemBase & problem, SystemBase & system);
+
+  /**
+   * Callback to the NonLinearTimeIntegratorInterface called immediately after the
+   * residuals are computed in NonlinearSystem::computeResidual().
+   * The residual vector which is passed in to this function should
+   * be filled in by the user with the _Re_time and _Re_non_time
+   * vectors in a way that makes sense for the particular
+   * TimeIntegration method.
+   */
+  virtual void postResidual(NumericVector<Number> & /*residual*/) {}
+
+  /**
+   * Returns the tag for the nodal multiplication factor for the residual calculation of the udot
+   * term.
+   *
+   * By default, this tag will be associated with udot.
+   */
+  TagID uDotFactorTag() const { return _u_dot_factor_tag; }
```

Comment:
should we error on those? seems it s only really implemented by CentralDifference

Reply:
I'll try to add these at another level and check if that triggers app failures.

Reply:
I think they ll need to be declared still on TimeIntegrator still
I dont see the code for the actual functionality in TimeIntegrator. Do you? Or is it all in CentralDiff?

Reply:
No, it is called in Inertia-related objects. But I assume those only work with the given time integrator. The interface has to stay, but with every other integrator we can just error out.

Reply:
Pushing soon.

Reply:
Never mind, it is used in `FEProblemBase::addTimeIntegrator` too. I might be able to solve that though.

Reply:
Nah, these tags might be pointing to udot or udotdot if you are not doing the central difference. I'll leave it as is.

Reply:
It is this one:
```
for (auto & nl : _nl)
  {
    nl->addDotVectors();

    auto tag_udot = nl->getTimeIntegrators()[0]->uDotFactorTag();
    if (!nl->hasVector(tag_udot))
      nl->associateVectorToTag(*nl->solutionUDot(), tag_udot);
    auto tag_udotdot = nl->getTimeIntegrators()[0]->uDotDotFactorTag();
    if (!nl->hasVector(tag_udotdot) && uDotDotRequested())
      nl->associateVectorToTag(*nl->solutionUDotDot(), tag_udotdot);
  }
```

Reply:
ok it's just extra not-always-necessary tagging. sounds good
thanks for taking a look.